### ISSUE TITLE
Error:  Protocol "https" not supported or disabled in libcurl

### ISSUE THREAD

**Issue by RobertAnderson-0609** [aws/aws-sdk-cpp#1186]
### What platform/OS are you using?
ARM-IMX-61 arm_a9hf Linux BSP 4.1.15
uname -m = 
armv7l

### Which version of the SDK?
aws sdk 1.7.11
with 
libcurl/7.65.1 OpenSSL/1.0.2s zlib/1.2.8 nghttp2/1.39.1

curl -V =
curl 7.65.1 (arm-navico-linux-gnueabi) libcurl/7.65.1 OpenSSL/1.0.2s zlib/1.2.8 nghttp2/1.39.1
Release-Date: 2019-06-05
Protocols: dict file ftp ftps gopher http https imap imaps pop3 pop3s rtsp smb smbs smtp smtps telnet tftp
Features: AsynchDNS HTTP2 HTTPS-proxy IPv6 Largefile libz NTLM NTLM_WB SSL TLS-SRP UnixSockets

### What compiler are you using? what version?
x86 (32-bit) to arm_a9 
arm-navico-linux-gnueabi-gcc (GCC) 5.3.0

### What are your CMake arguments?

export CMAKE_PREFIX_PATH="${SYSROOT}/usr"

cmake .. \
    -DCMAKE_TOOLCHAIN_FILE="${SYSROOT}/Toolchain.cmake" \
    -DCMAKE_INSTALL_PREFIX="${PKG}/usr" \
    -DCMAKE_BUILD_TYPE=RelWithDebInfo \
    -DCURL_HAS_H2_EXITCODE=0 \
    -DCURL_HAS_H2_EXITCODE__TRYRUN_OUTPUT="" \
    -DCURL_HAS_TLS_PROXY_EXITCODE=0 \
    -DCURL_HAS_TLS_PROXY_EXITCODE__TRYRUN_OUTPUT="" \
    -DBUILD_SHARED_LIBS="OFF" \
    -DBUILD_DEPS="OFF" \
    -DENABLE_CURL_LOGGING="ON" \
    -DBUILD_ONLY="core;s3;transfer" 

// AWS C Common

export CMAKE_PREFIX_PATH="${SYSROOT}/usr"

cmake .. \
    -DCMAKE_TOOLCHAIN_FILE="${SYSROOT}/Toolchain.cmake" \
    -DCMAKE_INSTALL_PREFIX="${PKG}/usr" \
    -DCMAKE_BUILD_TYPE=RelWithDebInfo

// AWS Checksums

export CMAKE_PREFIX_PATH="${SYSROOT}/usr"

cmake .. \
    -DCMAKE_TOOLCHAIN_FILE="${SYSROOT}/Toolchain.cmake" \
    -DCMAKE_INSTALL_PREFIX="${PKG}/usr" \
    -DCMAKE_BUILD_TYPE=RelWithDebInfo

// AWS C Event Stream

export CMAKE_PREFIX_PATH="${SYSROOT}/usr"

cmake .. \
    -DCMAKE_TOOLCHAIN_FILE="${SYSROOT}/Toolchain.cmake" \
    -DCMAKE_INSTALL_PREFIX="${PKG}/usr" \
    -DCMAKE_BUILD_TYPE=RelWithDebInfo \
    -DCMAKE_MODULE_PATH="${SYSROOT}/usr/lib/cmake"

// Contents of the Toolchain.cmake file

==> /Toolchain.cmake <==
set(CMAKE_SYSTEM_PROCESSOR arm_a9hf)
set(TOOLCHAIN_HAS_GLIB True)
set(TOOLCHAIN_HAS_GSTREAMER_1.0 True)
set(TOOLCHAIN_HAS_GSTREAMER_1.0_IMX_COMMUNITY True)
set(TOOLCHAIN_HAS_ALSA True)
set(TOOLCHAIN_HAS_LIBUSB True)
set(TOOLCHAIN_HAS_SOCKETCAN True)
set(TOOLCHAIN_HAS_OPENSSL True)
set(TOOLCHAIN_HAS_NAVICOFB True)
set(TOOLCHAIN_HAS_GPU True)

//  These should be maintained in the fsl-gpu package.
add_compile_definitions(LINUX=1 EGL_API_FB=1)

set(CMAKE_SYSTEM_NAME Linux)
set(CMAKE_SYSTEM_VERSION 1)
// Careful: CMAKE_CURRENT_LIST_DIR has caveats
set(CMAKE_SYSROOT "/usr/local/arm_a9")
set(CMAKE_PREFIX_PATH ${CMAKE_PREFIX_PATH} ${CMAKE_SYSROOT}/usr/lib/qt)

find_program(CCACHE_PROGRAM ccache)
if(CCACHE_PROGRAM)
set_property(GLOBAL PROPERTY RULE_LAUNCH_COMPILE "${CCACHE_PROGRAM}")
endif()

// Specify the cross compiler
set(CMAKE_C_COMPILER ${CMAKE_SYSROOT}/bin/arm-navico-linux-gnueabi-gcc)
set(CMAKE_CXX_COMPILER ${CMAKE_SYSROOT}/bin/arm-navico-linux-gnueabi-g++)
set(CMAKE_OBJCOPY ${CMAKE_SYSROOT}/bin/arm-navico-linux-gnueabi-objcopy)
set(CMAKE_CXX_COMPILER_ID "GNU")
set(CMAKE_CXX_FLAGS "-march=armv7-a -mtune=cortex-a9 -mfpu=neon -mfloat-abi=hard")

// QT_USE_QSTRINGBUILDER, BOOST_SP_USE_PTHREADS and the forced cache
// setting of CMAKE_BUILD_TYPE should probably be maintained in NOS, not
// the toolchain.

add_compile_definitions(QT_USE_QSTRINGBUILDER BOOST_SP_USE_PTHREADS)

if(NOT CMAKE_BUILD_TYPE)
set(CMAKE_BUILD_TYPE RelWithDebInfo CACHE STRING "Options: Debug Release RelWithDebInfo MinSizeRel" FORCE)
endif(NOT CMAKE_BUILD_TYPE)

### Can you provide a TRACE level log? (sanitize any sensitive information)

DEBUG] 2019-07-04 12:01:16.807 AWSAuthV4Signer [2297607168] Signed Headers value:content-length;content-type;host;x-amz-content-sha256;x-amz-date;x-amz-meta-brand;x-amz-meta-model;x-amz-me
ta-serialnumber;x-amz-meta-time;x-amz-meta-version
[DEBUG] 2019-07-04 12:01:16.807 AWSAuthV4Signer [2297607168] Canonical Request String: PUT
/***filename***

content-length:1177502
content-type:binary/octet-stream
host:***bucket***.s3.amazonaws.com
x-amz-content-sha256:UNSIGNED-PAYLOAD
x-amz-date:20190704T120116Z
x-amz-meta-brand:***brand***
x-amz-meta-model:***model***
x-amz-meta-serialnumber:***sn***
x-amz-meta-time:2019-07-04T12:00:43Z
x-amz-meta-version:1

content-length;content-type;host;x-amz-content-sha256;x-amz-date;x-amz-meta-brand;x-amz-meta-model;x-amz-meta-serialnumber;x-amz-meta-time;x-amz-meta-version
UNSIGNED-PAYLOAD
[DEBUG] 2019-07-04 12:01:16.807 AWSAuthV4Signer [2297607168] Final String to sign: AWS4-HMAC-SHA256
20190704T120116Z
20190704/us-east-1/s3/aws4_request
ab947873daa83b746dae8b2e979fea34378eafd816f749c09b33a23e97f7e928
[DEBUG] 2019-07-04 12:01:16.807 AWSAuthV4Signer [2297607168] Final computed signing hash: a26bd7b5cc9f175e693a590da9bd628a02a034bd8891e9d906ad542887e46df4
[DEBUG] 2019-07-04 12:01:16.808 AWSAuthV4Signer [2297607168] Signing request with: AWS4-HMAC-SHA256 Credential=***accesskey***/20190704/us-east-1/s3/aws4_request, SignedHeaders=content
-length;content-type;host;x-amz-content-sha256;x-amz-date;x-amz-meta-brand;x-amz-meta-model;x-amz-meta-serialnumber;x-amz-meta-time;x-amz-meta-version, Signature=a26bd7b5cc9f175e693a590da9b
d628a02a034bd8891e9d906ad542887e46df4
[DEBUG] 2019-07-04 12:01:16.808 AWSClient [2297607168] Request Successfully signed
[TRACE] 2019-07-04 12:01:16.808 CurlHttpClient [2297607168] Making request to https://***bucket***.s3.amazonaws.com/***file***
[TRACE] 2019-07-04 12:01:16.808 CurlHttpClient [2297607168] Including headers:
[TRACE] 2019-07-04 12:01:16.808 CurlHttpClient [2297607168] authorization: AWS4-HMAC-SHA256 Credential=***accesskey***/20190704/us-east-1/s3/aws4_request, SignedHeaders=content-length;
content-type;host;x-amz-content-sha256;x-amz-date;x-amz-meta-brand;x-amz-meta-model;x-amz-meta-serialnumber;x-amz-meta-time;x-amz-meta-version, Signature=a26bd7b5cc9f175e693a590da9bd628a02a
034bd8891e9d906ad542887e46df4
[TRACE] 2019-07-04 12:01:16.808 CurlHttpClient [2297607168] content-length: 1177502
[TRACE] 2019-07-04 12:01:16.808 CurlHttpClient [2297607168] content-type: binary/octet-stream
[TRACE] 2019-07-04 12:01:16.808 CurlHttpClient [2297607168] host: ***bucket***.s3.amazonaws.com
[TRACE] 2019-07-04 12:01:16.808 CurlHttpClient [2297607168] user-agent: aws-sdk-cpp/1.7.132 Linux/4.1.15 armv7l GCC/5.3.0
[TRACE] 2019-07-04 12:01:16.808 CurlHttpClient [2297607168] x-amz-content-sha256: UNSIGNED-PAYLOAD
[TRACE] 2019-07-04 12:01:16.809 CurlHttpClient [2297607168] x-amz-date: 20190704T120116Z
[TRACE] 2019-07-04 12:01:16.809 CurlHttpClient [2297607168] x-amz-meta-brand: ***brand***
[TRACE] 2019-07-04 12:01:16.809 CurlHttpClient [2297607168] x-amz-meta-model: ***model***
[TRACE] 2019-07-04 12:01:16.809 CurlHttpClient [2297607168] x-amz-meta-serialnumber: ***sn***
[TRACE] 2019-07-04 12:01:16.809 CurlHttpClient [2297607168] x-amz-meta-time: 2019-07-04T12:00:43Z
[TRACE] 2019-07-04 12:01:16.809 CurlHttpClient [2297607168] x-amz-meta-version: 1
[DEBUG] 2019-07-04 12:01:16.809 CurlHandleContainer [2297607168] Attempting to acquire curl connection.
[INFO] 2019-07-04 12:01:16.809 CurlHandleContainer [2297607168] Connection has been released. Continuing.
[DEBUG] 2019-07-04 12:01:16.809 CurlHandleContainer [2297607168] Returning connection handle 0x965738f0
[DEBUG] 2019-07-04 12:01:16.809 CurlHttpClient [2297607168] Obtained connection handle 0x965738f0
[DEBUG] 2019-07-04 12:01:16.810 CURL [2297607168] (Text) Protocol "https" not supported or disabled in libcurl

[DEBUG] 2019-07-04 12:01:16.810 CURL [2297607168] (Text) Closing connection -1

[ERROR] 2019-07-04 12:01:16.810 CurlHttpClient [2297607168] Curl returned error code 1 - Unsupported protocol
[DEBUG] 2019-07-04 12:01:16.810 CurlHandleContainer [2297607168] Releasing curl handle 0x965738f0
[DEBUG] 2019-07-04 12:01:16.810 CurlHandleContainer [2297607168] Notified waiting threads.
[DEBUG] 2019-07-04 12:01:16.810 AWSClient [2297607168] Request returned error. Attempting to generate appropriate error codes from response
[ERROR] 2019-07-04 12:01:16.810 AWSClient [2297607168] HTTP response code: -1
Exception name:
Error message: Unable to connect to endpoint

/////////////////////////////////
Code
/////////////////////////////////
``
    Aws::SDKOptions options;
    options.loggingOptions.logLevel = Aws::Utils::Logging::LogLevel::Trace;
    Aws::InitAPI(options);

    Aws::Client::ClientConfiguration clientConfig;

    clientConfig.region = "us-east-1";
    {
        auto s3Client = Aws::MakeShared<Aws::S3::S3Client>(
            ALLOC_TAG,
            Aws::Auth::AWSCredentials(m_AccesKeyId, m_SecretAccessKey),
            clientConfig);
        {
            auto executor = Aws::MakeShared<Aws::Utils::Threading::PooledThreadExecutor>(ALLOC_TAG, 4);
            Aws::Transfer::TransferManagerConfiguration transferConfig(executor.get());
            transferConfig.s3Client = s3Client;

            transferConfig.transferStatusUpdatedCallback =
                [this](const Aws::Transfer::TransferManager*, const std::shared_ptr<const Aws::Transfer::TransferHandle>& handle) {

                    const Aws::Transfer::TransferStatus status = handle->GetStatus();
                    switch (status)
                    {
                    case Aws::Transfer::TransferStatus::COMPLETED:
                    {
                        OnUploadComplete();
                    }
                    break;

                    case Aws::Transfer::TransferStatus::CANCELED:
                    case Aws::Transfer::TransferStatus::FAILED:
                    case Aws::Transfer::TransferStatus::ABORTED:
                    {
                        int error = static_cast<int>(handle->GetLastError().GetResponseCode());
                        OnUploadError(error);
                    }
                    break;

                    case Aws::Transfer::TransferStatus::EXACT_OBJECT_ALREADY_EXISTS:
                    case Aws::Transfer::TransferStatus::NOT_STARTED:
                    case Aws::Transfer::TransferStatus::IN_PROGRESS:
                        break;
                    }
                };

            transferConfig.uploadProgressCallback =
                [this](const Aws::Transfer::TransferManager*, const std::shared_ptr<const Aws::Transfer::TransferHandle>& handle) {
                    OnUploadProgress(handle->GetBytesTransferred());
                };

            {
                std::shared_ptr<Aws::Transfer::TransferManager> transferManager = Aws::Transfer::TransferManager::Create(transferConfig);
                {

                    const std::shared_ptr<Aws::Transfer::TransferHandle> transferHandle = transferManager->UploadFile(
                        m_FilePath.toStdString().c_str(),   // fileName
                        m_BucketName.toStdString().c_str(), // bucketName
                        m_Name.toStdString().c_str(),       // keyName
                        "binary/octet-stream",              // contentType
                        awsMetaData);                       // metaData

                    transferHandle->WaitUntilFinished();
                }
            }
        }
    }

    Aws::ShutdownAPI(options);
}
``

**Comment by RobertAnderson-0609** [comment]
Changing the scheme to http, I was able to upload a file successfully to s3 bucket.
Is anyone able to provide assistance with what might be wrong or how to debug what is happening in curl in more detail.

**Comment by singku** [comment]
When compiling sdk, what's the curl version found from the cmake output?

**Comment by RobertAnderson-0609** [comment]
This from the CMakeCache.txt file for aws-sdk-cpp
```
//Details about finding CURL
FIND_PACKAGE_MESSAGE_DETAILS_CURL:INTERNAL=[/usr/local/arm_a9/usr/lib/libcurl.so][/usr/local/arm_a9/usr/include][v7.65.1()]
//Details about finding OpenSSL
FIND_PACKAGE_MESSAGE_DETAILS_OpenSSL:INTERNAL=[/usr/local/arm_a9/usr/lib/libcrypto.so][/usr/local/arm_a9/usr/include][c ][v1.0.2s()]
//Details about finding ZLIB
FIND_PACKAGE_MESSAGE_DETAILS_ZLIB:INTERNAL=[/usr/local/arm_a9/usr/lib/libz.so][/usr/local/arm_a9/usr/include][v1.2.8()]
```

**Comment by singku** [comment]
Based on curl's FAQ https://curl.haxx.se/docs/faq.html
Https is not supported usually when SSL is disabled or not found during build CURL.
But based on you `curl -v` output and the CMake output, they are the same version.
 Could you please also run `ldd /path/to/curl binary` see if the curl library is the same as the one found in CMake when building SDK?

**Comment by RobertAnderson-0609** [comment]
The dates and file sizes on the target device are consistent with the stripped versions of the files that our build system is creating.
```
# ldd /usr/bin/***app***

libcurl.so.4 => /usr/lib/libcurl.so.4 (0xb12c1000)

libssl.so.1.0.0 => /usr/lib/libssl.so.1.0.0 (0xb125f000)

# ll /usr/lib/libcurl*
lrwxrwxrwx    1 root     root            16 Jul  4 01:16 /usr/lib/libcurl.so -> libcurl.so.4.5.0*
lrwxrwxrwx    1 root     root            16 Jul  4 01:16 /usr/lib/libcurl.so.4 -> libcurl.so.4.5.0*
-rwxr-xr-x    1 root     root        368712 Jul  3 22:11 /usr/lib/libcurl.so.4.5.0*

# ll /usr/lib/libssl*
lrwxrwxrwx    1 root     root            15 Jul  4 01:16 /usr/lib/libssl.so -> libssl.so.1.0.0*
-r-xr-xr-x    1 root     root        351332 Jul  3 21:58 /usr/lib/libssl.so.1.0.0*
```

**Comment by singku** [comment]
[/usr/local/arm_a9/usr/lib/libcurl.so] is the one found by cmake to compile sdk, which is not the one from above ldd cmd

**Comment by RobertAnderson-0609** [comment]
For clarity on the build machine 
/usr/local/arm_a9/ is the CMAKE_SYSROOT so it's removed from the full path to get /usr/lib/libcurl.so
and libcurl.so and libcurl.so.4 are symlinks to the real library libcurl.so.4.5.0 

On the target device we have installed a stripped version of libcurl.so.4.5.0 with the same symlinks.
We have confirmed end to end the file sizes and checksums of the library and there are no differences.

Is there anything else we can do we can do to increase the trace output to help identify what is going on?
